File: poly2rec.py
import os
import shutil
import cv2
from datetime import datetime
import numpy as np
import logging

class Point:

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __lt__(self, other):

        if self.y != other.y:
            return self.y < other.y
        else:
            return self.x < other.x

    def __gt__(self, other):
        return self.x > other.x and self.y > other.y

    def __str__(self):
        return f"Point({self.x}, {self.y})"


class Rect:

    def __init__(self, x, y, w, h):
        self.bl = Point(x, y)
        self.tr = Point(x + w, y + h)
        self.w = w
        self.h = h

    def __str__(self):
        return f"bl: {self.bl} tr: {self.tr}  w: {self.w}  h: {self.h}"


def is_equal(a, b, e):
    return abs(a - b) < e


def updatePolygon(polygon, P_bl, P_br, P_tl):

    up_left = Point(P_bl.x, P_tl.y)
    up_right = Point(P_br.x, P_tl.y)
    insert_up_left, insert_up_right = True, True

    if up_right in polygon:
        polygon.remove(up_right)
        insert_up_right = False

    if up_left in polygon:
        polygon.remove(up_left)
        insert_up_left = False

    if P_bl in polygon:
        polygon.remove(P_bl)

    if P_br in polygon:
        polygon.remove(P_br)

    if insert_up_left:
        polygon.append(up_left)

    if insert_up_right:
        polygon.append(up_right)

    return polygon


def find_points_triangle(polygon):
    polygon_cp = polygon.copy()

    # bottom_left (bl), bottom_right (br), top_left (tl)
    P_bl, P_br, P_tl = None, None, None
    sorted_polygon = sorted(polygon, reverse=False)
    P_bl = sorted_polygon[0]

    the_rest_polygon = []
    for point in polygon_cp:
        if P_bl != point:
            if P_bl.x <= point.x:
                the_rest_polygon.append(point)

    P_br = min(sorted(the_rest_polygon, reverse=False))

    the_rest_polygon = []
    for point in polygon:
        if point != P_bl and point != P_br:
            if point.x == P_bl.x or point.y == P_bl.y:
                the_rest_polygon.append(point)

    P_tl = min(the_rest_polygon, key=lambda k: [k.x, k.y])

    return P_bl, P_br, P_tl


def poly2rec(polygon):
    rectangles = []
    while len(polygon) > 0:
        P_bl, P_br, P_tl = find_points_triangle(polygon)

        # print( "P_bl: ", P_bl, " P_br: ", P_br, "P_tl: ", P_tl)
        rect = Rect(P_bl.x, P_bl.y,
                    P_br.x - P_bl.x, P_tl.y - P_bl.y)

        rectangles.append(rect)

        updatePolygon(polygon, P_bl, P_br, P_tl)

    return rectangles

def info():
    logging.info("im here")
    
def main():

    polygon = [

        # CASE 1
        # Point(0, 1),
        # Point(0, 2),
        # Point(2, 2),
        # Point(2, 0),
        # Point(1, 0),
        # Point(1, 1)

        # CASE 2
        # Point(0, 1),
        # Point(0, 2),
        # Point(3, 2),
        # Point(3, 1),
        # Point(2, 1),
        # Point(2, 0),
        # Point(1, 0),
        # Point(1, 1)

        # CASE 3
        # Point(44352, 0),
        # Point(44352,18997),
        # Point(60034,18997),
        # Point(60034,24623),
        # Point(68048,24623),
        # Point(68048,36968),
        # Point(45240,36968),
        # Point(45240,40008),
        # Point(5614,40008),
        # Point(5614,26366),
        # Point(0,26366),
        # Point(0,1440),
        # Point(1440,1440),
        # Point(1440,0)

        # CASE 4
        # Point(0,0),
        # Point(0,12328),
        # Point(19032,12328),
        # Point(19032,2745),
        # Point(11948,2745),
        # Point(11948,0)

        # CASE 5
        # Point(22176,0),
        # Point(22176,27578),
        # Point(5614,27578),
        # Point(5614,26366),
        # Point(0,26366),
        # Point(0,1440),
        # Point(1440,1440),
        # Point(1440,0)

        # CASE 6
        Point(0,240),
        Point(0,740),
        Point(740,740),
        Point(740,0),
        Point(300,0),
        Point(300,200),
        Point(200,200),
        Point(200,240)
    ]

    # get rectangles
    polygon_cp = polygon.copy()
    rectangles = poly2rec(polygon)

    # visualize 
    save_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                 f"logs/poly2rec_{datetime.now().strftime('%Y-%m-%d')}")
    if not os.path.exists(save_path):
        os.makedirs(save_path)
    else:
        if os.path.isdir(save_path):
            shutil.rmtree(save_path)  # remove dir and all contains
            os.makedirs(save_path)
        else:
            raise ValueError("file {} is not a file or dir.".format(save_path))
        
    # logging
    logging.root.name = f"poly2rec"
    logging.basicConfig(
        level=logging.DEBUG,
        format="[%(levelname)-2s] %(message)s",
        handlers=[
            logging.FileHandler(
                os.path.join(save_path, f"poly2rec.log"),
                mode='w'
            ),
            logging.StreamHandler()
        ]
    )

    points_X = [] # store x coodinates
    points_Y = [] # store y coordinates
    

    for point in polygon_cp:
        points_X.append(point.x) # Xs 
       
        points_Y.append(point.y)  # Ys
    
    # width, height
    width = int(max(points_X))
    height = int(max(points_X))

    img = 255 * np.ones((height, width), dtype=np.uint8)
    
    pts = []
    for point in polygon_cp:
        pts.append([point.x, point.y])

    
    # draw the polygon with black colors
    cv2.fillPoly(img, pts=[np.array(pts)], color=(0, 0, 0))

    # save orignial polygon 
    image_file = os.path.join(save_path, 'polygon.png')
    cv2.imwrite(image_file, img)

    # draw rectangles
    for rec in rectangles:

        # print rectangles (bottom_left, top_right, width, height)
        logging.info(f"bottom_left ({rec.bl.x},{rec.bl.y}) top_right ({rec.tr.x}, {rec.tr.y}), width: {rec.w}, height: {rec.h}")

        cv2.rectangle(img, (rec.bl.x, rec.bl.y), (rec.bl.x + rec.w, rec.bl.y + rec.h), (255, 255, 255), 1)

    # save the polygon with rectanges
    image_file = os.path.join(save_path, 'polygon_with_rectangles.png')
    cv2.imwrite(image_file, img)



if __name__ == '__main__':
    main()

File: poly2rec.cpp

#include <cstdio>
#include <vector>
#include <list>
#include <tuple>
#include <iterator>
#include <algorithm>
#include <iostream>
#include "coordinate.hpp"

using namespace std;


template <typename T, typename const_iterator>
void poly2rec(const const_iterator& first, const const_iterator& last, std::vector< Rect<T> >& result);


int main()
{
    vector< Coor<int> > polygon;
    vector< Rect<int> > result;

    polygon.emplace_back(0, 1);
    polygon.emplace_back(0, 2);
    polygon.emplace_back(2, 2);
    polygon.emplace_back(2, 0);
    polygon.emplace_back(1, 0);
    polygon.emplace_back(1, 1);

    poly2rec(polygon.cbegin(), polygon.cend(), result);

    cout << "Rectangle list:\n";
    int i = 0;
    for (const auto& rect : result) {
        i++;
        cout << "Rectangle " << i << ": bl" << rect.getBL() << " " << "tr" << rect.getTR() << "\n";
    }

    return 0;
}

// rectangle: bottom_left (bl), top_left (tl), top_right (tr), bottom_right (br)
// return coordinates of a triangle for generating a rectangle: bottom_left (bl), bottom_right (br), top_left (tl)
template <typename const_iterator>
static auto findCoorTriangle(const const_iterator& first, const const_iterator& last) -> std::tuple<const_iterator, const_iterator, const_iterator>
{
    // find bottom_left (P_bl)
    const auto P_bl = std::min_element(first, last, [](const auto& lhs, const auto& rhs) -> bool {
                if (lhs.getY() != rhs.getY()) { return (lhs.getY() < rhs.getY()); }
                else { return (lhs.getX() < rhs.getX()); }
            });

    // find bottom_right P_br except P_bl
    const auto P_br = std::min_element(first, last, [&P_bl](const auto& lhs, const auto& rhs) -> bool {
                // skip the element P_bl
                if (lhs == (*P_bl)) { return false; }
                else if (rhs == (*P_bl)) { return true; }
                // find the smallest element
                else if (lhs.getY() != rhs.getY()) { return (lhs.getY() < rhs.getY()); }
                else { return (lhs.getX() < rhs.getX()); }
            });

    // find top_left P_tl
    const auto P_tl = std::min_element(first, last, [&P_bl, &P_br](const auto& lhs, const auto& rhs) -> bool {
                // skip the element (x, y) out of the range:
                // P_bl.getX() <= x < P_br.getX() && P_bl.getY() < y
                if ((lhs.getY() <= P_bl->getY()) || (lhs.getX() < P_bl->getX()) || (lhs.getX() >= P_br->getX())) { return false; }
                else if ((rhs.getY() <= P_bl->getY()) || (rhs.getX() < P_bl->getX()) || (rhs.getX() >= P_br->getX())) { return true; }
                // find the smallest element
                else if (lhs.getY() != rhs.getY()) { return (lhs.getY() < rhs.getY()); }
                else { return (lhs.getX() < rhs.getX()); }
                return true;
            });
            
    cout << "Coordinates for a triangle in a rectangle" << "\n";
    cout << "   P_bl: (x,y): (" << P_bl->getX() << ", " << P_bl->getY() << ")\n";
    cout << "   P_br: (x,y): (" << P_br->getX() << ", " << P_br->getY() << ")\n";
    cout << "   P_tl: (x,y): (" << P_tl->getX() << ", " << P_tl->getY() << ")\n";
    cout << "\n";

    return {P_bl, P_br, P_tl};
}


template <typename T>
static void updatePolygon(std::list< Coor<T> >& polygon, const Coor<T>& P_bl, const Coor<T>& P_br, const Coor<T>& P_tl)
{
    auto iter = polygon.begin();
    const Coor<T> up_left(P_bl.getX(), P_tl.getY()), up_right(P_br.getX(), P_tl.getY());
    bool insert_up_left = true, insert_up_right = true;

    // Update Policy:
    // 1. If the coordinate exists in the point array, then remove it from the point array
    // 2. Otherwise, add it into the point array
    while (iter != polygon.end()) {
        if (*iter == P_bl)        { iter = polygon.erase(iter); }
        else if (*iter == P_br)   { iter = polygon.erase(iter); }
        else if (*iter == up_left)  { iter = polygon.erase(iter); insert_up_left = false; }
        else if (*iter == up_right)  { iter = polygon.erase(iter); insert_up_right = false; }
        else                    { ++iter; }
    }

    if (insert_up_left) { polygon.emplace_back(up_left); }
    if (insert_up_right) { polygon.emplace_back(up_right); }
}


template <typename T, typename const_iterator>
void poly2rec(const const_iterator& first, const const_iterator& last, std::vector< Rect<T> >& result)
{
    std::list< Coor<T> > polygon(first, last);
    
    result.clear();
    while (polygon.size() > 0) {
        const auto& coord_triangle = findCoorTriangle(polygon.cbegin(), polygon.cend());
        const auto P_bl = *(std::get<0>(coord_triangle));
        const auto P_br = *(std::get<1>(coord_triangle));
        const auto P_tl = *(std::get<2>(coord_triangle));

        // extract rectangle
        result.emplace_back(P_bl.getX(), P_bl.getY(), P_br.getX() - P_bl.getX(), P_tl.getY() - P_bl.getY());

        // update polygon (point array)
        updatePolygon(polygon, P_bl, P_br, P_tl);
    }
}
